File: client/java/src/main/java/io/armadaproject/ArmadaClient.java
package io.armadaproject;

import api.EventGrpc;
import api.EventOuterClass.EventStreamMessage;
import api.EventOuterClass.JobSetRequest;
import api.Health.HealthCheckResponse;
import api.Health.HealthCheckResponse.ServingStatus;
import api.Job.JobStatusRequest;
import api.Job.JobStatusResponse;
import api.JobsGrpc;
import api.SubmitGrpc;
import api.SubmitOuterClass.CancellationResult;
import api.SubmitOuterClass.JobCancelRequest;
import api.SubmitOuterClass.JobSubmitRequest;
import api.SubmitOuterClass.JobSubmitResponse;
import api.SubmitOuterClass.Queue;
import api.SubmitOuterClass.QueueGetRequest;
import api.SubmitOuterClass.StreamingQueueGetRequest;
import api.SubmitOuterClass.StreamingQueueMessage;
import com.google.protobuf.Empty;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Metadata;
import io.grpc.stub.MetadataUtils;
import io.grpc.stub.StreamObserver;
import java.util.Iterator;

public class ArmadaClient implements AutoCloseable {

  private final ManagedChannel channel;

  public ArmadaClient(String host, int port) {
    channel = ManagedChannelBuilder.forAddress(host, port).usePlaintext().build();
  }

  public ArmadaClient(String host, int port, String bearerToken) {
    channel = ManagedChannelBuilder
        .forAddress(host, port)
        .useTransportSecurity()
        .intercept(MetadataUtils.newAttachHeadersInterceptor(createAuthMetadata(bearerToken)))
        .build();
  }

  private Metadata createAuthMetadata(String token) {
    Metadata metadata = new Metadata();
    metadata.put(Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER),
        "Bearer " + token);
    return metadata;
  }

  public ServingStatus checkHealth() {
    SubmitGrpc.SubmitBlockingStub submitBlockingStub = SubmitGrpc.newBlockingStub(channel);
    HealthCheckResponse healthCheckResponse = submitBlockingStub.health(Empty.getDefaultInstance());
    return healthCheckResponse.getStatus();
  }

  public JobSubmitResponse submitJob(JobSubmitRequest jobSubmitRequest) {
    SubmitGrpc.SubmitBlockingStub submitBlockingStub = SubmitGrpc.newBlockingStub(channel);
    return submitBlockingStub.submitJobs(jobSubmitRequest);
  }

  public CancellationResult cancelJob(JobCancelRequest jobCancelRequest) {
    SubmitGrpc.SubmitBlockingStub submitBlockingStub = SubmitGrpc.newBlockingStub(channel);
    return submitBlockingStub.cancelJobs(jobCancelRequest);
  }

  public Iterator<EventStreamMessage> getEvents(JobSetRequest jobSetRequest) {
    EventGrpc.EventBlockingStub eventBlockingStub = EventGrpc.newBlockingStub(channel);
    return eventBlockingStub.getJobSetEvents(jobSetRequest);
  }

  public void streamEvents(JobSetRequest jobSetRequest,
      StreamObserver<EventStreamMessage> streamObserver) {
    EventGrpc.EventStub eventStub = EventGrpc.newStub(channel);
    eventStub.getJobSetEvents(jobSetRequest, streamObserver);
  }

  public Queue getQueue(QueueGetRequest queueGetRequest) {
    SubmitGrpc.SubmitBlockingStub submitBlockingStub = SubmitGrpc.newBlockingStub(channel);
    return submitBlockingStub.getQueue(queueGetRequest);
  }

  public Iterator<StreamingQueueMessage> getQueueInfo(
      StreamingQueueGetRequest streamingQueueGetRequest) {
    SubmitGrpc.SubmitBlockingStub submitBlockingStub = SubmitGrpc.newBlockingStub(channel);
    return submitBlockingStub.getQueues(streamingQueueGetRequest);
  }

  public JobStatusResponse getJobStatus(JobStatusRequest jobStatusRequest) {
    JobsGrpc.JobsBlockingStub jobsBlockingStub = JobsGrpc.newBlockingStub(channel);
    return jobsBlockingStub.getJobStatus(jobStatusRequest);
  }

  @Override
  public void close() {
    channel.shutdown();
  }

}

File: client/scala/armada-scala-client/src/main/scala/io/armadaproject/armada/ArmadaClient.scala
package io.armadaproject.armada

import api.job.{JobStatusRequest, JobStatusResponse, JobsGrpc}
import api.event.EventGrpc
import api.submit.{
  CancellationResult,
  JobCancelRequest,
  JobSetCancelRequest,
  JobSubmitRequest,
  JobSubmitRequestItem,
  JobSubmitResponse,
  Queue,
  QueueDeleteRequest,
  QueueGetRequest,
  SubmitGrpc
}
import api.health.HealthCheckResponse
import com.google.protobuf.empty.Empty
import io.grpc.netty.NettyChannelBuilder
import io.grpc.stub.MetadataUtils
import io.grpc.{ManagedChannel, ManagedChannelBuilder, Metadata}

import scala.concurrent.Future

class ArmadaClient(channel: ManagedChannel) {
  def submitJobs(
      queue: String,
      jobSetId: String,
      jobRequestItems: Seq[JobSubmitRequestItem]
  ): JobSubmitResponse = {
    val blockingStub = SubmitGrpc.blockingStub(channel)
    blockingStub.submitJobs(JobSubmitRequest(queue, jobSetId, jobRequestItems))
  }

  def getJobStatus(jobId: String): JobStatusResponse = {
    val blockingStub = JobsGrpc.blockingStub(channel)
    blockingStub.getJobStatus(JobStatusRequest(jobIds = Seq(jobId)))
  }

  def cancelJobs(
      cancelReq: JobCancelRequest
  ): scala.concurrent.Future[CancellationResult] = {
    SubmitGrpc.stub(channel).cancelJobs(cancelReq)
  }

  def cancelJobSet(jobSetId: String): Future[Empty] = {
    SubmitGrpc.blockingStub(channel).cancelJobSet(JobSetCancelRequest(jobSetId))
    Future.successful(new Empty)
  }

  def eventHealth(): HealthCheckResponse.ServingStatus = {
    val blockingStub = EventGrpc.blockingStub(channel)
    blockingStub.health(Empty()).status
  }

  def submitHealth(): scala.concurrent.Future[HealthCheckResponse] = {
    SubmitGrpc.stub(channel).health(Empty())
  }

  def createQueue(name: String): Unit = {
    val blockingStub = SubmitGrpc.blockingStub(channel)
    val q = api.submit.Queue().withName(name).withPriorityFactor(1)
    blockingStub.createQueue(q)
  }

  def deleteQueue(name: String): Unit = {
    val qReq = QueueDeleteRequest(name)
    val blockingStub = SubmitGrpc.blockingStub(channel)
    blockingStub.deleteQueue(qReq)
  }

  def getQueue(name: String): Queue = {
    val qReq = QueueGetRequest(name)
    val blockingStub = SubmitGrpc.blockingStub(channel)
    blockingStub.getQueue(qReq)
  }
}

object ArmadaClient {
  def apply(channel: ManagedChannel): ArmadaClient = {
    new ArmadaClient(channel)
  }

  def apply(host: String, port: Int): ArmadaClient = {
    apply(host, port, None)
  }

  def apply(
      host: String,
      port: Int,
      bearerToken: Option[String]
  ): ArmadaClient = {
    if (bearerToken.isEmpty) {
      val channel =
        NettyChannelBuilder.forAddress(host, port).usePlaintext().build()
      ArmadaClient(channel)
    } else {
      val metadata = new Metadata()
      metadata.put(
        Metadata.Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER),
        "Bearer " + bearerToken.get
      )
      val channel =
        NettyChannelBuilder
          .forAddress(host, port)
          .useTransportSecurity()
          .intercept(MetadataUtils.newAttachHeadersInterceptor(metadata))
          .build
      ArmadaClient(channel)
    }
  }
}
